### PR TITLE
DAOS-2496 test: mdtest with POSIX api

### PR DESCRIPTION
**Pull request by saurabhtandan**
Adding capability to mdtest_small.py to be able to
run with POSIX api over dfuse.
Currently only using single client node.

### CODE REVIEW

File: src/tests/ftest/io/mdtest_small.yaml [hunk 1]
```diff
@@ -17,14 +17,18 @@ pool:
     createset:
         group: daos_server
     createsize:
-        scm_size: 300000000
+        scm_size: 1600000000
```
Comment (mchaarawi) [inline src/tests/ftest/io/mdtest_small.yaml]: this pool size seems pretty large..

File: src/tests/ftest/util/mdtest_test_base.py [hunk 2]
```diff
@@ -58,7 +63,24 @@ def setUp(self):
...
+            self.hostlist_clients = [self.hostlist_clients[0]]
+            self.hostfile_clients = write_host_file.write_host_file(
+                self.hostlist_clients, self.workdir,
+                self.hostfile_clients_slots)
+
+    def tearDown(self):
+        """Tear down each test case."""
+        try:
+            if self.dfuse is not None:
+                self.dfuse.stop()
```
Comment (ashleypittman) [inline src/tests/ftest/util/mdtest_test_base.py]: It would be a lot nicer if the Dfuse() object could call self.stop() from a __del__() handler, it would avoid this code and a couple of similar function that were added in the initial commit.

File: src/tests/ftest/util/mdtest_test_base.py [hunk 3]
```diff
@@ -67,8 +89,70 @@ def execute_mdtest(self):
...
+        """Create a TestContainer object to be used to create container."""
+        # TO-DO: Enable container using TestContainer object,
+        # once DAOS-3355 is resolved.
+        # Get Container params
+        #self.container = TestContainer(self.pool)
+        #self.container.get_params(self)
+
+        # create container
+        # self.container.create()
+        env = Dfuse(self.hostlist_clients, self.tmp).get_default_env()
```
Comment (ashleypittman) [inline src/tests/ftest/util/mdtest_test_base.py]: I know this is also a copy of other code but it seems a odd way of doing things.  This line will create a Dfuse() object, read the device setting from the yaml file and close the object again, there's no need to create a Dfuse object to do this, can't it be done by a call to DaosContainer().create() instead?

Even if do you want to have this coded this way why not call .get_default_env() in self.dfuse?

File: src/tests/ftest/util/mdtest_test_base.py [hunk 4]
```diff
@@ -67,8 +89,70 @@ def execute_mdtest(self):
         # Create a pool
         self.pool.create()
 
-        # Run Mdtest
+    def create_cont(self):
```
Comment (ashleypittman) [inline src/tests/ftest/util/mdtest_test_base.py]: _create_cont() might be better here.

File: src/tests/ftest/util/mdtest_test_base.py [hunk 5]
```diff
@@ -67,8 +89,70 @@ def execute_mdtest(self):
...
+        self.dfuse.set_dfuse_cont_param(self.create_cont())
+
+        try:
+            # start dfuse
+            self.dfuse.run()
+        except CommandFailure as error:
+            self.log.error("Dfuse command %s failed on hosts %s",
+                           str(self.dfuse), str(NodeSet(self.dfuse.hosts)),
+                           exc_info=error)
+            self.fail("Test was expected to pass but it failed.\n")
```
Comment (ashleypittman) [inline src/tests/ftest/util/mdtest_test_base.py]: The text here isn't correct.

Is there any benefit in trying to catch this exception anyway?

File: src/tests/ftest/util/mdtest_test_base.py [hunk 6]
```diff
@@ -67,8 +89,70 @@ def execute_mdtest(self):
...
+                                         shell=True)
+            (output, err) = container.communicate()
+            self.log.info("Container created with UUID %s", output.split()[3])
+
+        except subprocess.CalledProcessError as err:
+            self.fail("Container create failed:{}".format(err))
+
+        return output.split()[3]
+
+    def start_dfuse(self):
```
Comment (ashleypittman) [inline src/tests/ftest/util/mdtest_test_base.py]: _start_dfuse()

File: src/tests/ftest/util/mdtest_test_base.py [hunk 7]
```diff
@@ -67,8 +89,70 @@ def execute_mdtest(self):
...
+            # start dfuse
+            self.dfuse.run()
+        except CommandFailure as error:
+            self.log.error("Dfuse command %s failed on hosts %s",
+                           str(self.dfuse), str(NodeSet(self.dfuse.hosts)),
+                           exc_info=error)
+            self.fail("Test was expected to pass but it failed.\n")
+
+
+    def execute_mdtest(self):
```
Comment (ashleypittman) [inline src/tests/ftest/util/mdtest_test_base.py]: Assuming this is the actual test it starts dfuse it's then stopped in the tearDown() outside of the test, so it would be better to stop dfuse from this function as well so shutdown errors could be reported properly as well.

File: src/tests/ftest/util/mdtest_test_base.py (hunk 3 above)
Comment (saurabhtandan) [reply src/tests/ftest/util/mdtest_test_base.py]: Yep, I am aware it is not one of the cleanest ways of implementing it. But this bit is just temporary. It should go away once the TO-DO bit mentioned above is implemented.
And regarding the second question, this is initiated here bcoz start_dfuse() needs it's own set of env which is slightly different that what container create need here. If I do it inside start_dfuse, it could interfere with those env and cause issues.

File: src/tests/ftest/util/mdtest_test_base.py (hunk 5 above)
Comment (saurabhtandan) [reply src/tests/ftest/util/mdtest_test_base.py]: Any suggestions for Text?
We need to catch exception in order to trigger test failure here using avocado's "self.fail()", otherwise the test will be reported with a status "ERROR" instead we want it to report the status "FAIL"

File: src/tests/ftest/util/mdtest_test_base.py (hunk 7 above)
Comment (saurabhtandan) [reply src/tests/ftest/util/mdtest_test_base.py]: Sure will do.

File: src/tests/ftest/util/mdtest_test_base.py (hunk 4 above)
Comment (saurabhtandan) [reply src/tests/ftest/util/mdtest_test_base.py]: Will do

File: src/tests/ftest/util/mdtest_test_base.py (hunk 6 above)
Comment (saurabhtandan) [reply src/tests/ftest/util/mdtest_test_base.py]: will do

File: src/tests/ftest/io/mdtest_small.yaml (hunk 1 above)
Comment (ashleypittman) [reply src/tests/ftest/io/mdtest_small.yaml]: That itself seems odd, I'd expect IOR to use more space than mdtest (although it depends on params obviously) and the same test over dfs/dfuse should use almost exactly the same.

File: src/tests/ftest/util/mdtest_test_base.py (hunk 3 above)
Comment (ashleypittman) [reply src/tests/ftest/util/mdtest_test_base.py]: I can't see a TO-DO above, can you point it out to me please?

get_default_env() just returns a dict, it doesn't modify anything so I don't see why you need a whole new object rather than the current one to call it.  This code isn't even calling dfuse is it?

File: src/tests/ftest/util/mdtest_test_base.py (hunk 5 above)
Comment (ashleypittman) [reply src/tests/ftest/util/mdtest_test_base.py]: I'm more used to Unittest where I thought any exception would cause a failure, not an error.

In which case self.fail('Unable to launch dfuse\n') would be more accurate.

File: src/tests/ftest/io/mdtest_small.yaml (hunk 1 above)
Comment (saurabhtandan) [reply src/tests/ftest/io/mdtest_small.yaml]: Reducing it will make the test fail. If this is a concern I can open a separate ticket to investigate this.

Comment (mchaarawi) [reply src/tests/ftest/io/mdtest_small.yaml]: i saw it first as 16g which i thought it was large. i see it's 1.6 which is fine.
but depending on the number of files and file size, you might be legitimately running out of space.

File: src/tests/ftest/io/mdtest_small.yaml [hunk 8]
```diff
@@ -38,7 +42,7 @@ mdtest:
     branch: !mux
         zero:
             branching_factor: 1
-            num_of_files_dirs: 1000
+            num_of_files_dirs: 1
```
Comment (mchaarawi) [inline src/tests/ftest/io/mdtest_small.yaml]: why did you change this to 1?
im not sure if im missing something in this yaml file, but that just creates a single empty file / dir from mdtest perspective, which really is close to not doing anything..

Comment (saurabhtandan) [reply src/tests/ftest/io/mdtest_small.yaml]: Was accidentally pushed as '1' made it back to 1000.

File: src/tests/ftest/io/mdtest_small.yaml (hunk 1 above)
Comment (ashleypittman) [reply src/tests/ftest/io/mdtest_small.yaml]: Regardless of the absolute size it's still five times bigger than the previous value, does this imply that mdtest over dfuse takes five times as much space?  As I understand it this is really about enabling essentially the same test over a different transport layer, the space used should remain the same so this is something to look into.

Yes please to a seperate ticket to investigate this.  Does either the mdtest-over-dvs or the mdtest-over-dfuse test log the amount of space used?

File: src/tests/ftest/util/mdtest_test_base.py (hunk 3 above)
Comment (ashleypittman) [reply src/tests/ftest/util/mdtest_test_base.py]: Sorry, I was wrong about the return type.

You haven't explained why this line can't self 'env = self.get_default_env()'
